packages/playwright-core/src/server/injected/selectorGenerator.ts

```diff
@@ -141,45 +142,48 @@ function generateSelectorFor(injectedScript: InjectedScript, targetElement: Elem
   return calculateCached(targetElement, true);
 }
 
-function buildCandidates(injectedScript: InjectedScript, element: Element): SelectorToken[] {
+function buildCandidates(element: Element): SelectorToken[] {
   const candidates: SelectorToken[] = [];
-  for (const attribute of ['data-testid', 'data-test-id', 'data-test']) {
-    if (element.getAttribute(attribute))
-      candidates.push({ engine: 'css', selector: `[${attribute}=${quoteAttributeValue(element.getAttribute(attribute)!)}]`, score: 1 });
+
+  if (element.getAttribute('data-testid'))
+    candidates.push({ engine: 'attr', selector: `[data-testid=${quoteAttributeValue(element.getAttribute('data-testid')!)}]`, score: 1 });
+
+  const ariaRole = getAriaRole(element);
+  if (ariaRole) {
+    const ariaName = getElementAccessibleName(element, false, new Map());
```

Comment:
Using a single `Map` as a cache during a single `generateSelector` call will be much faster. Accessible name is quite slow to compute.